yaf-src/airframe/doc/airdaemon.1
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "AIRDAEMON 1"
.TH AIRDAEMON 1 "20-Dec-2023" "2.15.0" "airframe"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
airdaemon \- Run a program as a daemon process, restarting it if it dies
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 5
\&    airdaemon   [\-\-retry RETRY_DELAY] [\-\-retry\-max RETRY_MAX_DELAY]
\&                [\-\-pidfile PID_FILE] [\-\-airdaemon\-pidfile PID_FILE]
\&                [\-\-log LOG_SPECIFIER] [\-\-loglevel LOG_LEVEL]
\&                [\-\-verbose] [\-\-version] [\-\-no\-daemon]
\&                \-\- PROGRAM [PROGRAM_ARGS]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBairdaemon\fR invokes a child process and ensures that it is restarted if it
encounters any errors.  The delay between child exit and restart can be
chosen, and can either be constant or exponentially increasing up to a
specified maximum.
.PP
Use of two dashes (\-\-) after all airdaemon command-line switches allows
\&\s-1PROGRAM_ARGS\s0 to be interpreted by the \s-1PROGRAM\s0 rather than \fBairdaemon\fR itself.
While they are not strictly required if you do not need to pass arguments to
\&\s-1PROGRAM,\s0 they should be used for consistency.
.SH "OPTIONS"
.IX Header "OPTIONS"
.SS "Daemon Options"
.IX Subsection "Daemon Options"
.IP "\fB\-\-retry\fR=\fI\s-1RETRY_DELAY\s0\fR" 4
.IX Item "--retry=RETRY_DELAY"
Wait \s-1RETRY_DELAY\s0 seconds after the child process exits abnormally until trying
to restart the child process.  See also \fB\-\-retry\-max\fR
.IP "\fB\-\-retry\-max\fR=\fI\s-1RETRY_MAX_DELAY\s0\fR" 4
.IX Item "--retry-max=RETRY_MAX_DELAY"
This parameter, combined with the \fB\-\-retry\fR parameter, facilitates truncated
binary exponential backoff of the retry logic.  The first retry attempt will
be delayed by the value of the \fB\-\-retry\fR parameter.  If the child process
exits abnormally within this time period, the delay is doubled.  Once the
child process survives longer than the value of the \fB\-\-retry\fR parameter, the
retry delay is reset to that value.
.Sp
For example, assume \fB\-\-retry\fR=30 and \fB\-\-retry\-max\fR=240.  The first time the
child process exits abnormally, \fBairdaemon\fR will wait 30 seconds before
invoking the child process again.  If it exits abnormally again within 30
seconds, \fBairdaemon\fR will wait 60 seconds and try again.  If the process
dies again within 30 seconds, the delay doubles again to 120.  The process
continues until the delay hits the maximum of 240.  Once the child process
survives for 30 seconds, the delay will be reset to the original retry delay
of 30.
.IP "\fB\-\-pidfile\fR=\fI\s-1PID_FILE\s0\fR" 4
.IX Item "--pidfile=PID_FILE"
Write the process identifier of the child process to \fI\s-1PID_FILE\s0\fR.
Each time the child process is restarted, the process identifier in this file
is updated. This option exists to facilitate the termination of the child
process by shutdown scripts. Note that if the child process terminates
normally during shutdown, airdaemon will terminate normally as well, so it
is usually sufficient just to kill the pid in this file at shutdown.
.IP "\fB\-\-airdaemon\-pidfile\fR=\fI\s-1PID_FILE\s0\fR" 4
.IX Item "--airdaemon-pidfile=PID_FILE"
Write the process identifier of the airdaemon process to \fI\s-1PID_FILE\s0\fR.
This option exists to facilitate the termination of the forked
airdaemon process by shutdown scripts.
.IP "\fB\-\-no\-daemon\fR" 4
.IX Item "--no-daemon"
Do not actually daemonize.  Mainly useful for testing/debugging.
.SS "Logging Options"
.IX Subsection "Logging Options"
These options are used to specify how log messages are routed.  airdaemon can
log to standard error, regular files, or the \s-1UNIX\s0 syslog facility.
.IP "\fB\-\-log\fR \fI\s-1LOG_SPECIFIER\s0\fR" 4
.IX Item "--log LOG_SPECIFIER"
Specifies destination for log messages. \fI\s-1LOG_SPECIFIER\s0\fR can be a \fBsyslog\fR\|(3)
facility name, the special value \fBstderr\fR for standard error, or the
\&\fIabsolute\fR path to a file for file logging. Standard error logging is only
available in \fB\-\-daemon\fR mode if \fB\-\-foreground\fR is present. The default log
specifier is \fBstderr\fR if available, \fBuser\fR otherwise.
.IP "\fB\-\-loglevel\fR \fI\s-1LOG_LEVEL\s0\fR" 4
.IX Item "--loglevel LOG_LEVEL"
Specify minimum level for logged messages. In increasing levels of verbosity,
the supported log levels are \fBquiet\fR, \fBerror\fR, \fBcritical\fR, \fBwarning\fR,
\&\fBmessage\fR, \fBinfo\fR, and \fBdebug\fR. The default logging level is \fBwarning\fR.
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
Equivalent to \fB\-\-loglevel debug\fR.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
If present, print version and copyright information to standard error and exit.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
The following will ping www.cert.org every 15 seconds until ping returns
sucess:
.PP
.Vb 1
\&    airdaemon \-\-retry 15 \-\- ping \-c 1 www.cert.org
.Ve
.PP
The same as above, except the delay will double until reaching a maximum of 15
minutes:
.PP
.Vb 1
\&    airdaemon \-\-retry 15 \-\-retry\-max 900 \-\- ping \-c 1 www.cert.org
.Ve
.SH "BUGS"
.IX Header "BUGS"
Known issues are listed in the \fB\s-1README\s0\fR file in the Airframe source
distribution. Note that airdaemon should be considered alpha-quality software;
not every concievable input and aggregation is exhaustively tested at each
release, and specific features may be completely untested. Please be mindful
of this before deploying airdaemon in production environments. Bug reports and
feature requests may be sent via email to <netsa\-help@cert.org>.
.SH "AUTHORS"
.IX Header "AUTHORS"
Tony Cebzanov and  Brian Trammell for the
\&\s-1CERT\s0 Network Situational Awareness
Group, http://www.cert.org/netsa
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBfiledaemon\fR\|(1)


yaf-src/NEWS
Version 2.15.0: 2023-12-21
==========================

Enhanced the deep packet inspection capabilities for SSH connections to include negotiated algorithms and HASSH hash.

Added the JA3 hash to the DPI for TLS connections.

Added support for reading VxLAN-encapsulated packets, Geneve-encapsulated packets, and Geneve-encapsulated VxLAN-encapsulated packets.

Fixed TLS certificate parsing to be more selective on which values are stored in the list of sslObjectType-sslObjectValue pairs.

Fixed a potential bug in the Shannon entropy calculation that may cause small differences in calculated values.

Version 3.0.0.alpha4: 2023-12-21
================================

Added the ability to decode VxLAN-encapsulated packets, Geneve-encapsulated packets, and Geneve-encapsulated VxLAN-encapsulated packets.

Added a --no-mpls option to disable export of MPLS labels when YAF has been built with MPLS support.

Fixed TLS certificate parsing to be more selective on which values are stored in the list of sslObjectType-sslObjectValue pairs.

Version 3.0.0.alpha3: 2023-07-18
================================

Changed DNS deep packet inspection to produce names and text records with escape codes for special characters (non-ASCII, non-printable, special whitespace, and label-internal dots in names).

Made DNS deep packet inspection more strict about parsing malformed DNS Resource Records across RR boundaries within the packet.

Enhanced the --time and --etime options of yafMeta2Pcap to accept a human-readable timestamp in addition to milliseconds.

Changed the destination of --version output to the standard output.

Changed yaf to only export the fingerprint-related elements (firstPacketBanner, etc) when the --fpexport option is given.  (Requires YAF to be built with --enable-fpexporter.)

Changed yaf to only export the p0f-related elements (osName, etc) when the --p0fprint option is given.  (Requires YAF to be built with --with-p0f.)

Fixed a crash in YAF that occurs when it is built with GLib 2.75.3 or newer.

Version 2.14.0: 2023-03-23
==========================

Changed DNS deep packet inspection to produce names and text records with escape codes for special characters (non-ASCII, non-printable, special whitespace, and label-internal dots in names).

Made DNS deep packet inspection more strict about parsing malformed DNS Resource Records across RR boundaries within the packet.

Changed destination of --version output to the standard output.

Fixed a crash in YAF that occurs when it is built with GLib 2.75.3 or newer.

Version 3.0.0.alpha2: 2023-02-09
================================

Enhanced the deep packet inspection capabilities for SSH connections to include negotiated algorithms and HASSH hash.

Added the JA3 hash to the DPI for TLS connections.

Made several changes to the yafDPIRules.conf file for applabels written as C plugins: Allow the user to disable the export of arbitrary DPI elements and SMTP headers. Allow a protocol to be specified. Moved the regex definitions from C to yafDPIRules.conf.

Increased the maximum payload that YAF may capture for performing DPI.

Fixed a potential bug in the Shannon entropy calculation that may cause small differences in calculated values.

Version 2.13.0: 2023-02-09
==========================

Added ability for yaf to limit payload export to a named set of applabels.

Increased the maximum payload that YAF may capture for performing DPI.

Added support for recent releases of nDPI.

Added yaf.init to the list of installed files.

Stopped export of full flow template that is never used for data records.

Fixed minor bug in --version where Compact IPv4 support always reported NO.

Fixed bugs in regular expressions for nntpResponseRegex and smtpURLRegex.

Version 3.0.0.alpha1: 2022-02-28
================================

Merged the configuration files yafApplabelRules.conf and yafDPIRules.conf into a single file written in Lua. Previous versions of those files will not work with this version of yaf.

Changed Deep Packet Inspection (DPI) support to be compiled into yaf when requested by configure; it is no longer a plug-in. Run configure with --enable-dpi to enable the capability; run yaf with --dpi to use it. Specifying --dpi enables application labeling; it is no longer necessary to explicitly specify --applabel when enabling DPI.

Changed yaf to export metadata about information elements and templates by default: both as compile-time and run-time options. To disable on an invocation, run yaf with the --no-element-metadata and/or --no-template-metadata switches. To disable support entirely, pass --disable-metadata-export to configure. (Note that super_mediator-2.0.0 works best with template metadata enabled.)

Updated yaf to use the enhanced template metadata available in libfixbuf-3.0.0.  This allows yaf to declare that it only uses some templates within sub-records (that is, within a subTemplateList or subTemplateMultiList).  The metadata also describes the information element yaf uses in its basicLists.

Added the yaf command line option --payload-applabel-select to enable exporting payload data for only selected appLabel values.

Updated the regular expressions used for application-labeling.

Changed numerous aspects of the DPI data.

Updated, rearranged, and fixed bugs in SMTP DPI.

Added fields for more DNSSEC values and fixed other bugs in DNS DPI.

Renamed the configure option --enable-p0fprinter to --with-p0f.

Renamed the configure option --enable-ndpi to --with-ndpi.

Fixed bugs in POP3 DPI.

Removed support for the Spread toolkit.

Removed support for the popt options parser.

Updated fixbuf requirement to libfixbuf-3.0.0.

Version 2.12.2: 2021-10-14
==========================

Added new protocols to the yafAppLabelRules.conf file and updated several regular expressions.

Changed the regexes used by the SMTP DPI plugin and improved capture when multiple messages appear in a single SMTP session.


yaf-src/airframe/src/airdaemon.c
/*
 *  Copyright 2007-2023 Carnegie Mellon University
 *  See license information in LICENSE.txt.
 */
/*
 *  airdaemon.c
 *  Keeps a child process running.
 *
 *  ------------------------------------------------------------------------
 *  Authors: Tony Cebzanov
 *  ------------------------------------------------------------------------
 *  @DISTRIBUTION_STATEMENT_BEGIN@
 *  YAF 2.15.0
 *
 *  Copyright 2023 Carnegie Mellon University.
 *
 *  NO WARRANTY. THIS CARNEGIE MELLON UNIVERSITY AND SOFTWARE ENGINEERING
 *  INSTITUTE MATERIAL IS FURNISHED ON AN "AS-IS" BASIS. CARNEGIE MELLON
 *  UNIVERSITY MAKES NO WARRANTIES OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
 *  AS TO ANY MATTER INCLUDING, BUT NOT LIMITED TO, WARRANTY OF FITNESS FOR
 *  PURPOSE OR MERCHANTABILITY, EXCLUSIVITY, OR RESULTS OBTAINED FROM USE OF
 *  THE MATERIAL. CARNEGIE MELLON UNIVERSITY DOES NOT MAKE ANY WARRANTY OF
 *  ANY KIND WITH RESPECT TO FREEDOM FROM PATENT, TRADEMARK, OR COPYRIGHT
 *  INFRINGEMENT.
 *
 *  Licensed under a GNU GPL 2.0-style license, please see LICENSE.txt or
 *  contact permission@sei.cmu.edu for full terms.
 *
 *  [DISTRIBUTION STATEMENT A] This material has been approved for public
 *  release and unlimited distribution.  Please see Copyright notice for
 *  non-US Government use and distribution.
 *
 *  GOVERNMENT PURPOSE RIGHTS - Software and Software Documentation
 *  Contract No.: FA8702-15-D-0002
 *  Contractor Name: Carnegie Mellon University
 *  Contractor Address: 4500 Fifth Avenue, Pittsburgh, PA 15213
 *
 *  The Government's rights to use, modify, reproduce, release, perform,
 *  display, or disclose this software are restricted by paragraph (b)(2) of
 *  the Rights in Noncommercial Computer Software and Noncommercial Computer
 *  Software Documentation clause contained in the above identified
 *  contract. No restrictions apply after the expiration date shown
 *  above. Any reproduction of the software or portions thereof marked with
 *  this legend must also reproduce the markings.
 *
 *  This Software includes and/or makes use of Third-Party Software each
 *  subject to its own license.
 *
 *  DM23-2313
 *  @DISTRIBUTION_STATEMENT_END@
 *  ------------------------------------------------------------------------
 */

#define _AIRFRAME_SOURCE_
#include <airframe/autoinc.h>
#include <airframe/airopt.h>
#include <airframe/logconfig.h>

static uint32_t       ad_retry_min        = 30;
static uint32_t       ad_retry_max        = 0;
static gboolean       ad_nodaemon         = FALSE;
static pid_t          ad_pid              = 0;
static char          *ad_pidfile          = NULL;
static char          *ad_cpidfile         = NULL;


static AirOptionEntry ad_options[]  = {
    AF_OPTION( "retry", 'r', 0, AF_OPT_TYPE_INT, &ad_retry_min,
               "Retry delay in seconds", "sec" ),
    AF_OPTION( "retry-max", 'R', 0, AF_OPT_TYPE_INT, &ad_retry_max,
               "Retry delay maximum in seconds", NULL ),
    AF_OPTION( "pidfile", 'P', 0, AF_OPT_TYPE_STRING, &ad_cpidfile,
               "A filename to write the child process pid to", NULL ),
    AF_OPTION( "airdaemon-pidfile", 'A', 0, AF_OPT_TYPE_STRING, &ad_pidfile,
               "A filename to write airdaemon's pid to", NULL ),
    AF_OPTION( "no-daemon", (char)0, 0, AF_OPT_TYPE_NONE, &ad_nodaemon,
               "do not daemonize", NULL ),
    AF_OPTION_END
};

typedef struct _ad_child_data {
    GMainLoop  *loop;
    gboolean   *done;
} ad_child_data_t;

static void
parse_options(
    int   *argc,
    char **argv[])
{
    AirOptionCtx *aoctx = NULL;

    aoctx = air_option_context_new("", argc, argv, ad_options);
    logc_add_option_group(aoctx, "airdaemon", VERSION);

    air_option_context_set_help_enabled(aoctx);

    air_option_context_parse(aoctx);
}


static void
on_child_exit(
    GPid       child_pid,
    gint       status,
    gpointer   data)
{
    GMainLoop *loop = ((ad_child_data_t *)data)->loop;
    gboolean  *done = ((ad_child_data_t *)data)->done;

    g_message("pid %lu exited with status %d", (gulong)child_pid, status);

#ifdef G_OS_UNIX
    if (WIFEXITED(status)) {
        if (WEXITSTATUS(status) == EXIT_SUCCESS) {
            g_debug("pid %lu returned success", (gulong)child_pid);
            *done = TRUE;
        } else {
            g_warning("pid %lu returned error status %d", (gulong)child_pid,
                      WEXITSTATUS(status));
        }
    } else if (WIFSIGNALED(status)) {
        g_critical("pid %lu terminated with signal %d\n",
                   (gulong)child_pid, WTERMSIG(status));
    } else {
        g_critical("pid %lu terminated", (gulong)child_pid);
    }
#endif /* G_OS_UNIX */
    g_spawn_close_pid(child_pid);
    g_main_loop_quit(loop);
}


static gboolean
daemonize(
    void)
{
    /* fork */
    if (fork()) {exit(0);}

    /* dissociate from controlling terminal */
    if (setsid() < 0) {
        g_critical("setsid() failed: %s", strerror(errno));
        return FALSE;
    }

    /* redirect stdio */
    freopen("/dev/null", "r", stdin);
    freopen("/dev/null", "w", stdout);
    freopen("/dev/null", "w", stderr);

    ad_pid = getpid();
    if (ad_pidfile) {
        FILE *pidfile = fopen(ad_pidfile, "w");
        if (!pidfile) {
            g_critical("could not write pidfile");
            goto end;
        }
        fprintf(pidfile, "%d\n", ad_pid);
        fclose(pidfile);
    }
  end:

    return TRUE;
}


int
main(
    int    argc,
    char  *argv[])
{
    int        i;
    gboolean   done = FALSE;
    GError    *err = NULL;
    uint32_t   delay;
    GTimer    *uptimer  = NULL;
    gdouble    elapsed_time;

    GMainLoop *loop;

    GPtrArray *child_args     = NULL;

    /* parse options */
    parse_options(&argc, &argv);

    /* set up logging */
    if (!logc_setup(&err)) {
        air_opterr("%s", err->message);
    }

    if (ad_retry_max && (ad_retry_min > ad_retry_max) ) {
        air_opterr("--retry value (%d) cannot exceed --retry-max value (%d) ",
                   ad_retry_min,
                   ad_retry_max);
    }
    delay = ad_retry_min;

    child_args = g_ptr_array_sized_new(64);
    for (i = 1; i < argc; i++) {
        /* Double dash indicates end of airdaemon's arguments */
        if (!strncmp(argv[i], "--", strlen(argv[i])) ) {
            continue;
        }
        g_ptr_array_add(child_args, g_strdup(argv[i]));
    }
    g_ptr_array_add(child_args, NULL);

    loop = g_main_loop_new(NULL, FALSE);

    /* Options check out; daemonize */
    if (!ad_nodaemon) {
        if (!daemonize()) {
            goto end;
        }
    }

    uptimer = g_timer_new();

    while (!done) {
        GPid            child_pid;
        char          **child_envp            = {NULL};
        GError         *child_err             = NULL;
        ad_child_data_t child_data;

        if (!g_spawn_async_with_pipes(".",
                                      (gchar **)child_args->pdata,
                                      child_envp,
                                      G_SPAWN_SEARCH_PATH |
                                      G_SPAWN_DO_NOT_REAP_CHILD,
                                      NULL,
                                      NULL,
                                      &child_pid,
                                      NULL,
                                      NULL,
                                      NULL,
                                      &child_err))
        {
            g_error("error spawning process: %s",
                    (child_err && child_err->message ? child_err->
                     message : "unknown error"));
        }

        g_timer_start(uptimer);

        /* Write child pid if requested */
        if (ad_cpidfile) {
            FILE *cpidfile = fopen(ad_cpidfile, "w");
            if (!cpidfile) {
                g_critical("could not write pidfile");
                goto end;
            }
            fprintf(cpidfile, "%d\n", child_pid);
            fclose(cpidfile);
        }

        /* Watch for process exit status */
        child_data.loop = loop;
        child_data.done = &done;

        g_child_watch_add(child_pid, on_child_exit, &child_data);
        g_main_loop_run(loop);

        g_timer_stop(uptimer);
        elapsed_time = g_timer_elapsed(uptimer, NULL);

        if (done) {
            g_debug("done");
        } else {
            if (ad_retry_max && (elapsed_time >= ad_retry_min) ) {
                g_debug("child survived for %fs, resetting delay",
                        elapsed_time);
                delay = ad_retry_min;
            }
            g_debug("child exited abnormally, sleeping for %d seconds", delay);
            sleep(delay);
            if (ad_retry_max) {
                if (2 * delay <= ad_retry_max) {
                    delay *= 2;
                } else {
                    delay = ad_retry_max;
                }
            }
        }
    }

  end:

    return 0;
}
